exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -216,90 +218,138 @@ fn create_java_keypair<'a>(
-    let num_services = env.get_array_length(services)?;
-    for i in 0..num_services {
-        let service_obj = env.get_object_array_element(services, i)?;
+    let num_artifacts = env.get_array_length(service_artifact_specs)?;
+    for i in 0..num_artifacts {
+        let artifact_spec_obj =
+            env.auto_local(env.get_object_array_element(service_artifact_specs, i)?);
 
-        let artifact_id = get_string_value(env, service_obj, "artifactId")?;
-        let artifact_filename = get_string_value(env, service_obj, "artifactFilename")?;
+        let artifact_id = to_artifact_id_str(get_field_as_string(
+            env,
+            artifact_spec_obj.as_obj(),
+            "artifactId",
+        )?);
+        let artifact_filename =
+            get_field_as_string(env, artifact_spec_obj.as_obj(), "artifactFilename")?;
         let service_specs_obj: jobjectArray = env
             .get_field(
-                service_obj,
+                artifact_spec_obj.as_obj(),
                 "serviceSpecs",
-                "[Lcom/exonum/binding/testkit/ServiceSpec;",
+                SERVICE_SPECS_FIELD_TYPE,
             )?
             .l()?
             .into_inner();
+        let configs = parse_service_specs(env, service_specs_obj, artifact_id, artifact_filename)?;
+        instance_configs.extend(configs);
+    }
+    Ok(instance_configs)
+}
 
-        let num_specs = env.get_array_length(service_specs_obj)?;
-
-        for k in 0..num_specs {
-            let service_spec = env.get_object_array_element(service_specs_obj, k)?;
-
-            let (service_id, service_name) = get_service_id_and_name(env, service_spec)?;
-            let config_params: jbyteArray = env
-                .get_field(service_spec, "configuration", "[B")?
-                .l()?
-                .into_inner();
-            let config = env.convert_byte_array(config_params)?;
+// Converts Java array of `ServiceSpec` instances into vector of `InstanceConfig` for specific artifact.
+fn parse_service_specs(
+    env: &JNIEnv,
+    specs_array: jobjectArray,
+    artifact_id: String,
+    artifact_filename: String,
+) -> JniResult<Vec<InstanceConfig>> {
+    let num_specs = env.get_array_length(specs_array)?;
 
-            let spec = InstanceSpec::new(service_id, service_name, &artifact_id)
-                .expect("Unable to create instance specification for service");
-            let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
-            instance_configs.push(cfg);
-        }
+    let mut instance_configs = vec![];
+    for i in 0..num_specs {
+        let service_spec = env.auto_local(env.get_object_array_element(specs_array, i)?);
+        let (spec, config) = parse_instance_spec(&env, service_spec.as_obj(), &artifact_id)?;
+        let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
+        instance_configs.push(cfg);
     }
+    env.delete_local_ref(specs_array.into())?;
```

Comment:
Because most probably we cannot use `AutoLocal` here - `AutoLocal` requires `JObject` and deletes the reference to it in `drop()`, but somewhere in between we have to convert that object into an array consuming it with `into_inner()`. In general, I agree that it will be better to re-write the whole thing to protobuf...

Reply:
If there is a deficiency in jni-rs — we can extend it.

The approach with manual management spanning multiple methods is not appropriate. Use frames if autolocals don't work.

Reply:
Tasks to rewrite this to protobuf: https://jira.bf.local/browse/ECR-3688, https://jira.bf.local/browse/ECR-3689